src/nci_adapter.c

```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+        prev->prot_info.size == curr->prot_info.size &&
+        !memcmp(prev->prot_info.bytes, curr->prot_info.bytes,
+            curr->prot_info.size);
+}
+
+static
+inline
+gboolean
+mode_param_match_poll_a_t2(
+    const NciModeParamPollA* prev,
+    const NciModeParamPollA* curr)
+{
+    gboolean partial_match =  mode_param_match_poll_a(prev, curr);
+    /* For tag type 2 logic is almost the same, but random UID has some
+    limitations: according to AN10927 Random UID RID should be handled
+    separately - single sized (4 bytes) starting with 0x08 */
+    if(prev->nfcid1_len == curr->nfcid1_len &&
+        curr->nfcid1_len == RANDOM_UID_SIZE &&
+        prev->nfcid1[0] == curr->nfcid1[0] &&
+        curr->nfcid1[0] == RANDOM_UID_START_BYTE) {
+        return partial_match;
+    } else {
+    /* otherwise UID should fully match */
+        return partial_match &&
+            prev->nfcid1_len == curr->nfcid1_len &&
+            !memcmp(prev->nfcid1, curr->nfcid1, curr->nfcid1_len);
+
+    }
+
+}
+
+static
+gboolean
+nci_adapter_info_mode_params_matches(
+    const NciAdapterIntfInfo* info,
+    const NciIntfActivationNtf* ntf)
+{
+    const NciModeParam* mode_param_prev = info->mode_param_parsed;
+    const NciModeParam* mode_param_current = ntf->mode_param;
+    /* mode params criteria depends on type of tag */
+    switch (ntf->mode) {
+    case NCI_MODE_PASSIVE_POLL_A:
+        switch (ntf->rf_intf) {
+        case NCI_RF_INTERFACE_FRAME:
+            /* Type 2 Tag */
+            if(mode_param_prev && mode_param_current) {
+                const NciModeParamPollA* poll_a_prev = &mode_param_prev->poll_a;
+                const NciModeParamPollA* poll_a_current =
+                    &mode_param_current->poll_a;
+
+                return mode_param_match_poll_a_t2(poll_a_prev, poll_a_current);
+            }
+            break;
+        case NCI_RF_INTERFACE_ISO_DEP:
+            /* ISO-DEP Type 4A */
+            if(mode_param_prev && mode_param_current) {
+                const NciModeParamPollA* poll_a_prev = &mode_param_prev->poll_a;
+                const NciModeParamPollA* poll_a_current =
+                    &mode_param_current->poll_a;
+
+                return mode_param_match_poll_a(poll_a_prev, poll_a_current);
+            }
+            break;
+        case NCI_RF_INTERFACE_NFCEE_DIRECT:
+        case NCI_RF_INTERFACE_NFC_DEP:
+        case NCI_RF_INTERFACE_PROPRIETARY:
+            break;
+        }
+        break;
+    case NCI_MODE_PASSIVE_POLL_B:
+        switch (ntf->rf_intf) {
+        case NCI_RF_INTERFACE_ISO_DEP:
+            /* ISO-DEP Type 4B */
+            if(mode_param_prev && mode_param_current) {
+                const NciModeParamPollB* poll_b_prev = &mode_param_prev->poll_b;
+                const NciModeParamPollB* poll_b_current =
+                    &mode_param_current->poll_b;
+
+                return mode_param_match_poll_b(poll_b_prev, poll_b_current);
+            }
+            break;
```

Comment:
```c
        case NCI_RF_INTERFACE_ISO_DEP: /* ISO-DEP Type 4B */
            return mode_param_match_poll_b(&mp1->poll_b, &mp2->poll_b);
```

Reply:
I have refactored that part of code. But unfortunately, parameters could be NULL, but I have moved that check one layer above (before switch